### ISSUE TITLE
[JIT] Optimization pass in profiling executor to fold away conditional branches

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#48089]
## 🚀 Feature

With profiling information, we should be able to remove conditional branches based on profiling information, including tensor ranks, device and boolean ivalue, e.t.c..

## Motivation

Conditional branches creates hierarchies that prevents us to exploit fusion opportunities, because currently fusion is limited to operations within the same block. With profiling executor, we should be able to fold away many conditional branches but unfortunately we don't have any passes to do that.

Two concrete examples that we are currently facing:

1. autodiff `dropout` contains many branches that stops us from fusing dropout with following operations. Which is a very common pattern.

https://github.com/pytorch/pytorch/blob/49f0e5dfeb64a928c6a2368dd5f86573b07d20fb/torch/csrc/jit/runtime/symbolic_script.cpp#L1083-L1111

2. python definition of `linear` layer has conditional eager fusion on gemm that kills the opportunity to fuse bias-add to following operations.

https://github.com/pytorch/pytorch/blob/49f0e5dfeb64a928c6a2368dd5f86573b07d20fb/torch/nn/functional.py#L1663-L1671

## Pitch

We can/should lift the checks of branching conditions to an earlier stage through exploiting profiling information. Things like tensor rank / device / training flag doesn't change often, if at all. This would greatly simplifies control flow and expose more fusion opportunities.

## Alternatives

A simpler WAR would be to remove the control flow in the graph. In the two examples mentioned above, we can:
1). For autodiff dropout, incorporate conditional variable as parameters of composite operation. e.g. a specialized `dropout_cuda` node that takes `train` as an argument;
2). For `torch.nn.linear`, maybe hide the eager mode fusion from JIT. Branch for `bias` is tricky and I don't have an easy solution.

## Additional context

Our investigation start from #45519.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
